Implement Plane methods

Robert7301201 opened the pull request:
Implement all of `Plane` methods supported by Godot.
Includes documentation and tests to make sure it matches Godot's outputs.
Resolves #612.

karroffel reviewed (commented):
Very nice, thanks for the PR!

I think it would make sense to use `approx_eq` for all the tests maybe? I know that normalising implementations of floating point arithmetic are actually deterministic for those common operations, but somehow I still feel we should give some wiggle room?

toasteater commented on gdnative-core/src/core_types/geom/plane.rs:
```diff
@@ -20,4 +21,262 @@ impl Plane {
+
+    /// Creates a new `Plane` from the ['Vector3'](./type.Vector3.html) normal and the distance from the origin.
+    #[inline]
+    pub fn new_with_normal(normal: Vector3, d: f32) -> Plane {
```
`from_normal` would be a name that fits more into Rust conventions.

toasteater commented on gdnative-core/src/core_types/geom/plane.rs:
```diff
@@ -20,4 +21,262 @@ impl Plane {
+    /// a, b, c are used for the normal ['Vector3'](./type.Vector3.html).
+    /// d is the distance from the origin.
+    #[inline]
+    pub fn new_with_reals(a: f32, b: f32, c: f32, d: f32) -> Plane {
```
`from_coordinates` would be a name that fits more into Rust conventions.

toasteater commented on gdnative-core/src/core_types/geom/plane.rs:
```diff
@@ -20,4 +21,262 @@ impl Plane {
+
+    /// Creates a new `Plane` from three [`Vector3`](./type.Vector3.html), given in clockwise order.
+    #[inline]
+    pub fn new_with_vectors(a: Vector3, b: Vector3, c: Vector3) -> Plane {
```
`from_points` would be a name that fits more into Rust conventions.

Also, how does this handle the case where all points are collinear, assuming that this is a constructor that creates a `Plane` that contains all three points? Maybe it could be documented?

toasteater commented on gdnative-core/src/core_types/geom/plane.rs:
```diff
@@ -20,4 +21,262 @@ impl Plane {
+    /// Creates a new `Plane` from the ['Vector3'](./type.Vector3.html) normal and the distance from the origin.
+    #[inline]
+    pub fn new_with_normal(normal: Vector3, d: f32) -> Plane {
+        Plane {
+            normal: normal,
+            d: d,
+        }
```
Can be simplified to `Plane { normal, d }`.

toasteater commented on gdnative-core/src/core_types/geom/plane.rs:
```diff
@@ -20,4 +21,262 @@ impl Plane {
+    pub fn normalize(mut self) -> Plane {
+        let l = self.normal.length();
+
+        if l == 0.0 {
+            self.normal = Vector3::new(0.0, 0.0, 0.0);
+            self.d = 0.0;
+            return self;
+        } else {
+            self.normal /= l;
+            self.d /= l;
+            return self;
+        }
```
Unnecessary `return`s here.

toasteater reviewed (commented):
Nice PR! I think the constructors can be renamed to fit better into Rust conventions. There are also a few style problems that might trigger clippy.

Use `approx_eq` for the tests as karroffel suggested and I think this looks good to me.

bors try

Robert7301201 commented:
I've implemented the requested changes. One issue though is that `from_points` will return `NaN`s if the points are all collinear. Do you want me to add error checking and return an `Option<Plane>` based on whether the points make a valid `Plane`? I added documentation so if we do decide to forgo error checking the user is at least aware of the risks.

toasteater commented:
I believe this should be checked in the library. `NaN`s are rarely considered in user logic (because it's extremely tedious to check for them everywhere), so it's easy for them to silently contaminate everything, resulting in problems that are hard to debug. An `Option<Plane>` should be fine.

Robert7301201 commented:
Done. I believe that's the only method that needs NaN checking. The others shouldn't produce NaNs unless provided with NaNs.

toasteater reviewed (approved):
Looks good to me. Thanks for the PR!

bors r+

toasteater commented:
Please run `cargo fmt` with the latest stable rustfmt.

toasteater commented:
bors r+
